api/src/main/java/io/druid/data/input/impl/prefetch/Fetcher.java

```diff
@@ -0,0 +1,324 @@
+ * See the javadoc of {@link PrefetchableTextFilesFirehoseFactory} for more details.
+ */
+public class Fetcher<ObjectType> implements Iterator<OpenedObject<ObjectType>>
+{
+  private static final Logger LOG = new Logger(Fetcher.class);
+  private static final String FETCH_FILE_PREFIX = "fetch-";
+
+  private final CacheManager<ObjectType> cacheManager;
+  private final List<ObjectType> objects;
+  private final ExecutorService fetchExecutor;
+  private final File temporaryDirectory;
+
+  // A roughly max size of total fetched objects, but the actual fetched size can be bigger. The reason is our current
+  // client implementations for cloud storages like s3 don't support range scan yet, so we must download the whole file
+  // at once. It's still possible for the size of cached/fetched data to not exceed these variables by estimating the
+  // after-fetch size, but it makes us consider the case when any files cannot be fetched due to their large size, which
+  // makes the implementation complicated.
+  private final long maxFetchCapacityBytes;
+  private final boolean prefetchEnabled;
+
+  private final long prefetchTriggerBytes;
+
+  // timeout for fetching an object from the remote site
+  private final long fetchTimeout;
+
+  // maximum retry for fetching an object from the remote site
+  private final int maxFetchRetry;
+
+  private final LinkedBlockingQueue<FetchedFile<ObjectType>> fetchedFiles = new LinkedBlockingQueue<>();
+
+  // Number of bytes of current fetched files.
+  // This is updated when a file is successfully fetched, a fetched file is deleted, or a fetched file is
+  // cached.
+  private final AtomicLong fetchedBytes = new AtomicLong(0);
+
+  private final ObjectOpenFunction<ObjectType> openObjectFunction;
+
+  private Future<Void> fetchFuture;
+
+  // nextFetchIndex indicates which object should be downloaded when fetch is triggered.
+  // This variable is always read by the same thread regardless of prefetch is enabled or not.
+  private int nextFetchIndex;
+
+  private int numRemainingObjects;
+
+  Fetcher(
+      CacheManager<ObjectType> cacheManager,
+      List<ObjectType> objects,
+      ExecutorService fetchExecutor,
+      File temporaryDirectory,
+      long maxFetchCapacityBytes,
+      long prefetchTriggerBytes,
+      long fetchTimeout,
+      int maxFetchRetry,
+      ObjectOpenFunction<ObjectType> openObjectFunction
+  )
+  {
+    this.cacheManager = cacheManager;
+    this.objects = objects;
+    this.fetchExecutor = fetchExecutor;
+    this.temporaryDirectory = temporaryDirectory;
+    this.maxFetchCapacityBytes = maxFetchCapacityBytes;
+    this.prefetchTriggerBytes = prefetchTriggerBytes;
+    this.fetchTimeout = fetchTimeout;
+    this.maxFetchRetry = maxFetchRetry;
+    this.openObjectFunction = openObjectFunction;
+
+    this.prefetchEnabled = maxFetchCapacityBytes > 0;
+    this.numRemainingObjects = objects.size();
+
+    if (cacheManager.isInitialized()) {
```

Comment:
IMO it's easier to understand and less error-prone when just one method on cacheManager is called in this small block: getFiles(), instead of three. And the other two are not needed at all.

Reply:
Fixed.